File: backend/app/api/v1/endpoints/tags.py
"""
Tag endpoints.

Provides REST API for managing tags and tag associations.
"""

from typing import List

from fastapi import APIRouter, Depends, Path, Query, status

from app.api.deps import get_tag_service
from app.domain.schemas.tag import (
    TableSyncTagAssociationCreate,
    TableSyncTagAssociationResponse,
    TableSyncTagsResponse,
    TagCreate,
    TagListResponse,
    TagResponse,
    TagSuggestionResponse,
)
from app.domain.services.tag import TagService

router = APIRouter()


@router.post(
    "",
    response_model=TagResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create tag",
    description="Create a new tag for smart grouping",
)
async def create_tag(
    tag_data: TagCreate, service: TagService = Depends(get_tag_service)
) -> TagResponse:
    """
    Create a new tag.

    Args:
        tag_data: Tag creation data
        service: Tag service instance

    Returns:
        Created tag
    """
    return service.create_tag(tag_data)


@router.get(
    "",
    response_model=TagListResponse,
    summary="List tags",
    description="Get a list of all tags with pagination",
)
async def list_tags(
    skip: int = Query(0, ge=0, description="Number of items to skip"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of items to return"),
    service: TagService = Depends(get_tag_service),
) -> TagListResponse:
    """
    List all tags with pagination.

    Args:
        skip: Number of tags to skip
        limit: Maximum number of tags to return
        service: Tag service instance

    Returns:
        List of tags with total count
    """
    return service.get_all_tags(skip=skip, limit=limit)


@router.get(
    "/search",
    response_model=TagSuggestionResponse,
    summary="Search tags",
    description="Search tags for autocomplete (partial match)",
)
async def search_tags(
    q: str = Query("", description="Search query"),
    limit: int = Query(10, ge=1, le=50, description="Maximum number of suggestions"),
    service: TagService = Depends(get_tag_service),
) -> TagSuggestionResponse:
    """
    Search tags for autocomplete.

    Args:
        q: Search query
        limit: Maximum number of suggestions
        service: Tag service instance

    Returns:
        List of suggested tags
    """
    return service.search_tags(query=q, limit=limit)


@router.get(
    "/{tag_id}",
    response_model=TagResponse,
    summary="Get tag",
    description="Get a specific tag by ID",
)
async def get_tag(
    tag_id: int = Path(..., ge=1, description="Tag ID"),
    service: TagService = Depends(get_tag_service),
) -> TagResponse:
    """
    Get tag by ID.

    Args:
        tag_id: Tag identifier
        service: Tag service instance

    Returns:
        Tag details
    """
    return service.get_tag(tag_id)


@router.delete(
    "/{tag_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete tag",
    description="Delete a tag by ID",
)
async def delete_tag(
    tag_id: int = Path(..., ge=1, description="Tag ID"),
    service: TagService = Depends(get_tag_service),
) -> None:
    """
    Delete a tag.

    Args:
        tag_id: Tag identifier
        service: Tag service instance
    """
    service.delete_tag(tag_id)


# Table Sync Tag Association Endpoints


@router.get(
    "/table-sync/{table_sync_id}",
    response_model=TableSyncTagsResponse,
    summary="Get tags for table sync",
    description="Get all tags associated with a pipeline destination table sync",
)
async def get_table_sync_tags(
    table_sync_id: int = Path(..., ge=1, description="Pipeline destination table sync ID"),
    service: TagService = Depends(get_tag_service),
) -> TableSyncTagsResponse:
    """
    Get all tags for a table sync.

    Args:
        table_sync_id: Pipeline destination table sync ID
        service: Tag service instance

    Returns:
        List of tags associated with the table sync
    """
    return service.get_tags_for_table_sync(table_sync_id)


@router.post(
    "/table-sync/{table_sync_id}",
    response_model=TableSyncTagAssociationResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Add tag to table sync",
    description="Add a tag to a pipeline destination table sync (creates tag if doesn't exist)",
)
async def add_tag_to_table_sync(
    table_sync_id: int = Path(..., ge=1, description="Pipeline destination table sync ID"),
    tag_data: TableSyncTagAssociationCreate = ...,
    service: TagService = Depends(get_tag_service),
) -> TableSyncTagAssociationResponse:
    """
    Add a tag to a table sync.

    If the tag doesn't exist, it will be created automatically.

    Args:
        table_sync_id: Pipeline destination table sync ID
        tag_data: Tag association data
        service: Tag service instance

    Returns:
        Created association
    """
    return service.add_tag_to_table_sync(table_sync_id, tag_data)


@router.delete(
    "/table-sync/{table_sync_id}/tags/{tag_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Remove tag from table sync",
    description="Remove a tag association from a pipeline destination table sync",
)
async def remove_tag_from_table_sync(
    table_sync_id: int = Path(..., ge=1, description="Pipeline destination table sync ID"),
    tag_id: int = Path(..., ge=1, description="Tag ID"),
    service: TagService = Depends(get_tag_service),
) -> None:
    """
    Remove a tag from a table sync.

    Args:
        table_sync_id: Pipeline destination table sync ID
        tag_id: Tag ID
        service: Tag service instance
    """
    service.remove_tag_from_table_sync(table_sync_id, tag_id)


@router.get(
    "/tag/{tag_id}/table-syncs",
    response_model=List[int],
    summary="Get table syncs by tag",
    description="Get all pipeline destination table sync IDs associated with a tag",
)
async def get_table_syncs_by_tag(
    tag_id: int = Path(..., ge=1, description="Tag ID"),
    service: TagService = Depends(get_tag_service),
) -> List[int]:
    """
    Get all table sync IDs associated with a tag.

    Args:
        tag_id: Tag ID
        service: Tag service instance

    Returns:
        List of table sync IDs
    """
    return service.get_table_syncs_by_tag(tag_id)

File: web/src/repo/tags.ts
import { AxiosResponse } from 'axios'
import { api } from './client'

export interface Tag {
  id: number
  tag: string
  created_at: string
  updated_at: string
}

export interface TagListResponse {
  tags: Tag[]
  total: number
}

export interface TagSuggestionResponse {
  suggestions: Tag[]
}

export interface TableSyncTagsResponse {
  table_sync_id: number
  tags: Tag[]
  total: number
}

export interface TableSyncTagAssociation {
  id: number
  pipelines_destination_table_sync_id: number
  tag_id: number
  tag_item: Tag
  created_at: string
  updated_at: string
}

export interface CreateTagRequest {
  tag: string
}

export interface AddTagToTableSyncRequest {
  tag: string
}

export const tagsRepo = {
  // Tag CRUD operations
  getAll: async (skip: number = 0, limit: number = 100): Promise<TagListResponse> => {
    const response: AxiosResponse<TagListResponse> = await api.get('/tags', {
      params: { skip, limit },
    })
    return response.data
  },

  create: async (data: CreateTagRequest): Promise<Tag> => {
    const response: AxiosResponse<Tag> = await api.post('/tags', data)
    return response.data
  },

  get: async (id: number): Promise<Tag> => {
    const response: AxiosResponse<Tag> = await api.get(`/tags/${id}`)
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/tags/${id}`)
  },

  search: async (query: string, limit: number = 10): Promise<TagSuggestionResponse> => {
    const response: AxiosResponse<TagSuggestionResponse> = await api.get('/tags/search', {
      params: { q: query, limit },
    })
    return response.data
  },

  // Table Sync Tag Association operations
  getTableSyncTags: async (tableSyncId: number): Promise<TableSyncTagsResponse> => {
    const response: AxiosResponse<TableSyncTagsResponse> = await api.get(
      `/tags/table-sync/${tableSyncId}`
    )
    return response.data
  },

  addTagToTableSync: async (
    tableSyncId: number,
    data: AddTagToTableSyncRequest
  ): Promise<TableSyncTagAssociation> => {
    const response: AxiosResponse<TableSyncTagAssociation> = await api.post(
      `/tags/table-sync/${tableSyncId}`,
      data
    )
    return response.data
  },

  removeTagFromTableSync: async (tableSyncId: number, tagId: number): Promise<void> => {
    await api.delete(`/tags/table-sync/${tableSyncId}/tags/${tagId}`)
  },

  getTableSyncsByTag: async (tagId: number): Promise<number[]> => {
    const response: AxiosResponse<number[]> = await api.get(`/tags/tag/${tagId}/table-syncs`)
    return response.data
  },
}
